pytential/symbolic/compiler.py

```diff
@@ -24,85 +24,125 @@
 """
 
 from dataclasses import dataclass
-from functools import reduce
-from typing import TYPE_CHECKING, Any
+from typing import TYPE_CHECKING
 
 import numpy as np
+from typing_extensions import override
 
-from pymbolic.primitives import Subscript, Variable, cse_scope
+import pymbolic.primitives as prim
+from pytools.obj_array import ObjectArray, ObjectArray1D, ShapeT, from_numpy
 
 from pytential.symbolic.mappers import CachedIdentityMapper, DependencyMapper
-from pytential.symbolic.primitives import DOFDescriptor, IntG, NamedIntermediateResult
+from pytential.symbolic.primitives import (
+    DOFDescriptor,
+    IntG,
+    NamedIntermediateResult,
+    QBXForcedLimit,
+)
 
 
 if TYPE_CHECKING:
     from collections.abc import Collection, Hashable, Iterator, Sequence, Set
 
+    from pymbolic.mapper.dependency import Dependency
     from pymbolic.typing import Expression
     from sumpy.kernel import Kernel
 
+    from pytential.collection import GeometryCollection
+
+__doc__ = """
+.. autoclass:: Statement
+.. autoclass:: Assign
+.. autoclass:: PotentialOutput
+.. autoclass:: ComputePotential
+
+.. autofunction:: dot_dataflow_graph
+
+.. autoclass:: Code
+.. autoclass:: OperatorCompiler
+"""
+
 
 # {{{ statements
 
 @dataclass(frozen=True, eq=False)
 class Statement:
     """
-    .. attribute:: names
-    .. attribute:: exprs
-    .. attribute:: priority
+    .. autoattribute:: names
+    .. autoattribute:: exprs
+    .. autoattribute:: priority
+
+    .. automethod:: get_assignees
+    .. automethod:: get_dependencies
     """
 
-    names: list[str]
-    exprs: list[Expression]
+    names: tuple[str, ...]
+    """Names of the variables used in this statement."""
+    exprs: tuple[Expression, ...]
```

Comment:
Made a few of these to requiring tuples, but hopefully it doesn't break anything (the classes themselves are already frozen).